Travis stages

kevnm67 opened the pull request:
## Pull request checklist

- [x] All tests pass. 
- [x] Demo project builds and runs.
- [x] I have resolved merge conflicts.
- [x] I have followed the [coding style](https://github.com/jessesquires/HowToContribute#style-guidelines) and [contributing guidelines](https://github.com/jessesquires/HowToContribute).

#### This fixes issue #
- #64 and builds on #129

## What's in this pull request?
- Updated travis to use build stages vs. matrix.
- [Added] Build carthage stage.

kevnm67 committed:
Added travis stages for tests and building the project

kevnm67 committed:
Added clean to xcodebuild test script

kevnm67 committed:
Updated build project script

- Fixed travis build error

kevnm67 commented:
@jessesquires Got this [working](https://github.com/kevnm67/JSQDataSourcesKit/pull/4) over the weekend.  I added #64.  Happy to break it out into a follow-up PR, though.

Let me know what you think and want any changes.

jessesquires commented on .travis.yml:
```diff
@@ -1,5 +1,6 @@
 os: osx
 osx_image: xcode10.1
+language: swift
```
do we need this? 

i think only specifying `os:` and `osx_image:` should work, right?

kevnm67 replied:
I go back and forth on this.  Sorry about that.  Ill remove it.

<img width="852" alt="image" src="https://user-images.githubusercontent.com/2734154/54678963-9d126600-4adc-11e9-95ad-ef80c9a0bfe6.png">


FYI, I found the [thread](https://github.com/travis-ci/docs-travis-ci-com/issues/922) about Travis CI treating ```objective-c``` and ```swift``` the same.

jessesquires replied:
💯

jessesquires commented on .travis.yml:
```diff
@@ -23,35 +24,58 @@ env:
+  include:
+    - &test
+      stage: test
+      name: iOS iPhone X 11.4
```
can we just use the OS name? 

so: `iOS 11.4`, `iOS 12.1`, `tvOS 12.1`

kevnm67 replied:
Sounds good.  Ill update those now.

jessesquires commented on .travis.yml:
```diff
@@ -23,35 +24,58 @@ env:
+    - &test
+      stage: test
+      name: iOS iPhone X 11.4
+      env: DESTINATION="OS=11.4,name=iPhone X" SDK="$IOS_SDK" BUILD_EXAMPLE="YES"
```
we can remove `BUILD_EXAMPLE="YES/NO"` from all of these 😊

kevnm67 replied:
Ha, of course!

jessesquires commented on .travis.yml:
```diff
@@ -23,35 +24,58 @@ env:
+      script:
+        - set -o pipefail
+        - xcodebuild -version -sdk
+        - xcodebuild clean test -project "$PROJECT" -scheme JSQDataSourcesKit -sdk "$SDK" -destination "$DESTINATION" -enableCodeCoverage YES -configuration Debug ONLY_ACTIVE_ARCH=NO CODE_SIGNING_REQUIRED=NO | bundle exec xcpretty -c;
```
now that I think about it, let's pull `JSQDataSourcesKit` out into a `$SCHEME` variable.

then this travis file will be easier to reuse in other projects 😄

jessesquires commented on .travis.yml:
```diff
@@ -23,35 +24,58 @@ env:
 
-- bundle exec danger
+    - &build
+      stage: build
```
can we call this `build examples` ?

kevnm67 replied:
Done

jessesquires commented on .travis.yml:
```diff
@@ -23,35 +24,58 @@ env:
+      osx_image: xcode10.1
+      name: Carthage
+      os: osx
+      script: carthage build --no-skip-current --configuration Release
```
nice! 👏 👏 👏

kevnm67 replied:
:) Thought I'd throw it in here... and ask for forgiveness if you requested to break it out 😺

jessesquires commented on .travis.yml:
```diff
@@ -23,35 +24,58 @@ env:
+  - test
+  - build
+  - lint
+  - carthage
```
💯

kevnm67 replied:
One question:  
Do you want to conditionally build any stage(s)?  (i.e. only run ```carthage``` if branch == master && !PR).

jessesquires replied:
meh, i think it's fine to always run. it shouldn't take long

(and i'd rather catch bugs early)

jessesquires reviewed (changes requested):
thanks @kevnm67 ! 

this is awesome 💯 🥇 

only a few minor comments

kevnm67 committed:
Renamed stages and updated env variable

- Created scheme env variable.
- Renamed jobs per OS being tested.
- Renamed build stage to build examples.
- Removed build example yes/no from test scripts.

jessesquires committed:
remove last BUILD_EXAMPLE

jessesquires committed:
consistent stage naming

jessesquires commented:
thanks again @kevnm67 ! 🌈 

this is fucking awesome

kevnm67 commented:
@jessesquires Thanks! I appreciate the kind words 😎
